File: packages/contracts/src/scientific.ts
import { z } from "zod"

import { NonEmptyTextSchema, Sha256Schema, UtcTimestampSchema, Uuid7Schema } from "./common"

export const MeasurementUnitSchema = z
  .strictObject({ quantity: NonEmptyTextSchema, ucum_code: NonEmptyTextSchema })
  .readonly()

export const CalibrationContextSchema = z
  .strictObject({
    method: NonEmptyTextSchema,
    reference: NonEmptyTextSchema,
    calibrated_at: UtcTimestampSchema,
    calibration_sha256: Sha256Schema,
  })
  .readonly()

export const ResearchContextSchema = z
  .strictObject({
    purpose: z.literal("research_only"),
    subject_scope: z.literal("non_clinical"),
    description: NonEmptyTextSchema,
  })
  .readonly()

export const ScientificInputSchema = z
  .strictObject({
    artifact_version_id: Uuid7Schema,
    kind: z.enum(["spectrum", "image", "table", "report"]),
    units: z.array(MeasurementUnitSchema).min(1).readonly(),
    calibration: CalibrationContextSchema,
    context: ResearchContextSchema,
  })
  .readonly()

export const ApprovedActionPlanRefSchema = z
  .strictObject({
    action_plan_id: Uuid7Schema,
    version: z.int().min(1),
    digest_sha256: Sha256Schema,
    approval_id: Uuid7Schema,
    approved_at: UtcTimestampSchema,
    immutable: z.literal(true),
  })
  .readonly()

export const ApprovedExecutionRefSchema = z
  .strictObject({
    execution_id: Uuid7Schema,
    action_plan_id: Uuid7Schema,
    approval_id: Uuid7Schema,
    status: z.literal("completed"),
    execution_sha256: Sha256Schema,
  })
  .readonly()

export type ScientificInput = z.infer<typeof ScientificInputSchema>
export type ApprovedActionPlanRef = z.infer<typeof ApprovedActionPlanRefSchema>
export type ApprovedExecutionRef = z.infer<typeof ApprovedExecutionRefSchema>

File: packages/contracts/python/science_workbench_contracts/scientific.py
from typing import Annotated, Literal

from pydantic import Field

from .common import NonEmptyText, Revision, Sha256, UtcTimestamp, Uuid7
from .task6_common import Task6ContractModel


class MeasurementUnit(Task6ContractModel):
    quantity: NonEmptyText
    ucum_code: NonEmptyText


class CalibrationContext(Task6ContractModel):
    method: NonEmptyText
    reference: NonEmptyText
    calibrated_at: UtcTimestamp
    calibration_sha256: Sha256


class ResearchContext(Task6ContractModel):
    purpose: Literal["research_only"]
    subject_scope: Literal["non_clinical"]
    description: NonEmptyText


class ScientificInput(Task6ContractModel):
    artifact_version_id: Uuid7
    kind: Literal["spectrum", "image", "table", "report"]
    units: Annotated[tuple[MeasurementUnit, ...], Field(min_length=1)]
    calibration: CalibrationContext
    context: ResearchContext


class ApprovedActionPlanRef(Task6ContractModel):
    action_plan_id: Uuid7
    version: Revision
    digest_sha256: Sha256
    approval_id: Uuid7
    approved_at: UtcTimestamp
    immutable: Literal[True]


class ApprovedExecutionRef(Task6ContractModel):
    execution_id: Uuid7
    action_plan_id: Uuid7
    approval_id: Uuid7
    status: Literal["completed"]
    execution_sha256: Sha256
